### ISSUE TITLE
Operations are slowed down by internal call to `closure`.

### ISSUE THREAD

**Issue by mjstahlberg** [ntessore/composition_stats#1]
I profiled a function computing the Aitchison distance using `inner` and `perturb_inv`. This was the result (ca. 55k calls):
![composition_stats](https://user-images.githubusercontent.com/3780206/136654400-4654bdb7-b68d-4854-83c5-339e7926ccb3.png)
Note that almost all of the clock time (32.8 out of 39.2 seconds) is spent inside `closure`. However my vectors are already compositional, so this is maybe unnecessary.

I see three non-exclusive options to improve this:

1. Improve the performance of `closure` (it seems rather slow for what I would imagine should be an all-positive check and subsequent division by `numpy.sum`).
2. Allow skipping automatic closure via a keyword argument to `inner` and others.
3. Add a specialized function to compute the Aitchison distance, performing fewer closure steps.

Thank you for this useful library!

**Comment by ntessore** [comment]
Thanks for pointing out this issue! I have pushed a commit to the branch which tries to address this:

1. In `closure()` compute the norm and check where its zero, instead of checking for rows of zeros.
2. A `closure(..., inplace=True)` option to perform the closure in place, which is used on intermediate arrays in other functions.
3. Optional parameters `normalized=True` for all functions which initially compute a closure, to skip that computation.

If you could install this* and see if it makes a difference, that would be much appreciated.

---

*) For example:
```
pip install git+https://github.com/ntessore/composition_stats.git#egg=composition_stats
```

**Comment by mjstahlberg** [comment]
Thank you for the quick response!

Without changing my code, being
```python
def aitchison_distance(comp1, comp2):
    """Compute the Aitchison distance."""
    difference = cs.perturb_inv(comp1, comp2)
    return cs.inner(difference, difference) ** 0.5
```
your change already gives a performance improvement in the order of 9%. Changing it to
```python
def aitchison_distance(comp1, comp2, normalized=True):
    """Compute the Aitchison distance."""
    difference = cs.perturb_inv(comp1, comp2, normalized=normalized)
    return cs.inner(difference, difference, normalized=True) ** 0.5
```
saves around 49% clock time total. Note that both `clr` (called by `inner`) and `perturb_inv` do still call `closure` internally; the number of calls to closure per call of `aitchison_distance` went down from 7 to 3 but it is still used. I'm not sure if these remaining three calls are intended?

**Comment by ntessore** [comment]
I have thought about his a bit more and decided to remove all internal calls to `closure()` entirely. It is much easier for the user to call `closure()` once on their input data before passing it to the functions than to always subsequently tell the functions whether the data is or isn't normalised.

This will mean a breaking change, and hence v2, but I suppose that in practice, it doesn't really make a lot of difference.

So let me know if the latest commit works for you. The remaining calls to `closure()` cannot be removed, because they are part of the transformations.

**Comment by mjstahlberg** [comment]
With my original implementation, we're talking about 72% of time saved now:

![aitchison_distance_2](https://user-images.githubusercontent.com/3780206/136685536-83f27352-5637-475b-b231-9ebc1f1e3beb.png)

Only a single call to `closure` per distance calculation remains. Therein is still some room for improvement (e.g. `np.atleast_2d` is responsible for 7% of the remaining time) but for my purposes this is fast enough.

**Comment by ntessore** [comment]
I have released composition_stats 2.0.0 with these changes. Thanks again for reporting this issue! I've only been using these functions for simple data visualisation, so it's great to see how much room for improvement there is when a package meets real applications for the first time.
